File: lib/features/spelling/domain/vi_spelling.dart
import 'package:unorm_dart/unorm_dart.dart' as unorm;

/// Vietnamese syllable decomposition result (matches Python dict structure)
class DecomposedSyllable {
  final String initial;
  final String nucleus;   // called "main" in Python
  final String ending;
  final String tone;
  final String original;
  final String spellString;
  final String ttsString;

  DecomposedSyllable({
    required this.initial,
    required this.nucleus,
    required this.ending,
    required this.tone,
    required this.original,
    required this.spellString,
    required this.ttsString,
  });

  @override
  String toString() => '[$original]: $spellString';
}

// ==================== CONSTANTS ====================

const List<String> INITIALS = [
  "ngh", "ng", "gh", "gi", "qu", "ch", "kh", "nh", "ph", "th", "tr",
  "b", "c", "d", "đ", "g", "h", "k", "l", "m", "n", "p", "q", "r",
  "s", "t", "v", "x",
];

const List<String> ENDINGS = ["ch", "nh", "ng", "c", "m", "n", "p", "t"];

const Map<String, Map<String, String>> _GI_SPECIAL_CASES = {
  "gì": {
    "initial": "gi",
    "main": "",
    "ending": "",
    "tone": "Huyền",
    "spell": "gi huyền gì",
    "tts": "gi huyền gì",
  },
  "gìn": {
    "initial": "gi",
    "main": "in",
    "ending": "",
    "tone": "Huyền",
    "spell": "gi in gin huyền gìn",
    "tts": "gi in gin huyền gìn",
  },
  "giếng": {
    "initial": "gi",
    "main": "iêng",
    "ending": "",
    "tone": "Sắc",
    "spell": "gi iêng giêng sắc giếng",
    "tts": "gi iêng giêng sắc giếng",
  },
  "giềng": {
    "initial": "gi",
    "main": "iêng",
    "ending": "",
    "tone": "Huyền",
    "spell": "gi iêng giêng huyền giềng",
    "tts": "gi iêng giêng huyền giềng",
  },
  "giết": {
    "initial": "gi",
    "main": "iết",
    "ending": "",
    "tone": "Sắc",
    "spell": "gi iết giết sắc giết",
    "tts": "gi iết giết sắc giết",
  },
  "giêng": {
    "initial": "gi",
    "main": "iêng",
    "ending": "",
    "tone": "Ngang",
    "spell": "gi iêng giêng",
    "tts": "gi iêng giêng",
  },
};

const Map<String, String> TONE_MAP = {
  "\u0300": "Huyền",
  "\u0309": "Hỏi",
  "\u0303": "Ngã",
  "\u0301": "Sắc",
  "\u0323": "Nặng",
};

const String VOWEL_ORDER = "aăâeêoôơiuưy";

// ==================== HELPER FUNCTIONS ====================

/// Add Vietnamese tone mark (exactly same logic as Python `add_tone`)
String addTone(String syllable, String tone) {
  if (syllable.isEmpty || tone == "" || tone == "0") return syllable;

  const toneSymbols = {
    "s": "\u0301", // sắc
    "f": "\u0300", // huyền
    "r": "\u0323", // nặng
    "x": "\u0309", // hỏi
    "j": "\u0303", // ngã
  };

  final combining = toneSymbols[tone.toLowerCase()] ?? "";
  if (combining.isEmpty) return syllable;

  int bestPos = -1;
  int bestPriority = VOWEL_ORDER.length;

  for (int i = 0; i < syllable.length; i++) {
    final lower = syllable[i].toLowerCase();
    final priority = VOWEL_ORDER.indexOf(lower);
    if (priority != -1 && priority < bestPriority) {
      bestPriority = priority;
      bestPos = i;
    }
  }

  if (bestPos == -1) return syllable;

  return syllable.substring(0, bestPos + 1) +
      combining +
      syllable.substring(bestPos + 1);
}

/// Main decomposition function - **Strictly follows Python logic**
DecomposedSyllable decomposeVietnameseSyllable(String word) {
  if (word.isEmpty) {
    return DecomposedSyllable(
      initial: "",
      nucleus: "",
      ending: "",
      tone: "Ngang",
      original: word,
      spellString: "",
      ttsString: "",
    );
  }

  final lowerWord = word.toLowerCase();

  // 1. Special cases for "gi"
  if (_GI_SPECIAL_CASES.containsKey(lowerWord)) {
    final sp = _GI_SPECIAL_CASES[lowerWord]!;
    return DecomposedSyllable(
      initial: sp["initial"]!,
      nucleus: sp["main"]!.isEmpty ? "(none)" : sp["main"]!,
      ending: sp["ending"]!.isEmpty ? "(none)" : sp["ending"]!,
      tone: sp["tone"]!,
      original: word,
      spellString: sp["spell"]!,
      ttsString: sp["tts"]!,
    );
  }

  // 2. Normalize to NFD and extract tone
  final normalized = unorm.nfd(lowerWord);

  String tone = "Ngang";
  for (final char in normalized.split('')) {
    if (TONE_MAP.containsKey(char)) {
      tone = TONE_MAP[char]!;
      break;
    }
  }

  // Remove tone marks
  final base = normalized.split('').where((c) => !TONE_MAP.containsKey(c)).join();

  // 3. Find initial consonant
  final chars = base.split('');
  String initial = "";
  int i = 0;

  final sortedInitials = List<String>.from(INITIALS)
    ..sort((a, b) => b.length.compareTo(a.length));

  for (final init in sortedInitials) {
    if (chars.length >= init.length &&
        chars.sublist(0, init.length).join() == init) {
      initial = init;
      i = init.length;
      break;
    }
  }

  List<String> remaining = chars.sublist(i);

  // 4. Find ending consonant
  String ending = "";
  if (remaining.isNotEmpty) {
    final sortedEndings = List<String>.from(ENDINGS)
      ..sort((a, b) => b.length.compareTo(a.length));

    for (final end in sortedEndings) {
      final endLen = end.length;
      if (remaining.length >= endLen &&
          remaining.sublist(remaining.length - endLen).join() == end) {
        ending = end;
        remaining = remaining.sublist(0, remaining.length - endLen);
        break;
      }
    }
  }

  // 5. Nucleus (main vowel)
  final main = remaining.join();
  final mainNfc = unorm.nfc(main);

  // 6. Build spelling parts (exactly as Python)
  List<String> parts = [];

  // Handle "qu" special case
  String currentMain = mainNfc;
  if (initial == "qu" && ending.isNotEmpty) {
    currentMain = "u" + mainNfc;
  }

  // 1. Spell main vowel letter by letter
  final mainEndingTemp = currentMain + ending;
  if (mainEndingTemp.length > 1) {
    for (final letter in currentMain.split('')) {
      parts.add(letter);
    }
  }

  // 2. Spell ending
  if (ending.isNotEmpty) {
    parts.add(ending);
  }

  // 3. Main + Ending together
  final mainEnding = currentMain + ending;
  if (mainEnding.length > 1) {
    parts.add(mainEnding);
  }

  // 4. Initial + Main_Ending
  String syllableNoTone = "";
  if (initial.isNotEmpty) {
    parts.add(initial);
    parts.add(mainEnding);

    // Fix double-u for "qu"
    String finalMainEnding = mainEnding;
    if (initial == "qu" && ending.isNotEmpty) {
      finalMainEnding = mainEnding.substring(1);
    }

    syllableNoTone = initial + finalMainEnding;
    if (syllableNoTone.isNotEmpty) {
      parts.add(syllableNoTone);
    }
  }

  // 5. Add tone name
  if (tone != "Ngang") {
    parts.add(tone.toLowerCase());
  }

  // 6. Final word with tone
  if (tone != "Ngang") {
    parts.add(word);
  }

  // 7. Single letter (alphabet case)
  if (word.length == 1) parts = [word];

  final spellString = parts.join(' ');

  // 8. Build TTS string
  List<String> ttsParts = List.from(parts);

  for (int idx = 0; idx < ttsParts.length; idx++) {
    String token = ttsParts[idx];

    // Add "ờ" to consonants
    if (INITIALS.contains(token) || ENDINGS.contains(token)) {
      if (token == "k") {
        ttsParts[idx] = "ca";
      } else if (token == "ngh") {
        ttsParts[idx] = "ngờ";
      } else if (token == "gh") {
        ttsParts[idx] = "gờ";
      } else if (token != "gi") {
        ttsParts[idx] = token + "ờ";
      }
    }

    // Special case "i" and "y"
    if ((token == "i" || token == "y") && (ttsParts.length > 1)){
      ttsParts[idx] += ttsParts[idx];
    }

    // Special case "q"
    if (token == "q"){
      ttsParts[idx] = "cu";
    }


    // Add sắc tone to closed syllables
    if ((token == mainEnding || token == syllableNoTone) &&
        ["ch", "c", "p", "t"].contains(ending)) {
      ttsParts[idx] = addTone(ttsParts[idx], "s");
    }

    // Special case: "qu" with stop ending
    if (initial == "qu" && ending.isNotEmpty) {
      if (ttsParts[idx].length > 1 &&
          ttsParts[idx].substring(1) == mainEnding &&
          ["ch", "c", "p", "t"].contains(ending)) {
        ttsParts[idx] = addTone(ttsParts[idx], "s");
      }
    }
  }

  final ttsString = ttsParts.join(' ');

  return DecomposedSyllable(
    initial: initial.isEmpty ? "(none)" : unorm.nfc(initial),
    nucleus: mainNfc.isEmpty ? "(none)" : unorm.nfc(mainNfc),
    ending: ending.isEmpty ? "(none)" : unorm.nfc(ending),
    tone: tone,
    original: unorm.nfc(word),
    spellString: unorm.nfc(spellString),
    ttsString: unorm.nfc(ttsString),
  );
}

/// Get full spelling for a sentence (now uses safer splitting)
DecomposedSyllable getSpellingForText(String text) {
  // Use the same regex style as Python, but more robust for Vietnamese
  final RegExp regex = RegExp(r'[\p{L}\p{N}]+', unicode: true);
  final words = regex
      .allMatches(text)
      .map((m) => m.group(0)!)
      .where((w) => w.trim().isNotEmpty)
      .toList();

  final ttsParts = <String>[];
  final spellParts = <String>[];
  for (final w in words) {
    final result = decomposeVietnameseSyllable(w);
    if (result.ttsString.isNotEmpty) {
      ttsParts.add(result.ttsString);
      ttsParts.add(""); //for a short break between each word.
      spellParts.add(result.spellString);
      spellParts.add(""); //for a short break between each word.
    }
  }
  return DecomposedSyllable(
    initial: "(none)",
    nucleus: "(none)",
    ending: "(none)",
    tone: "(none)",
    original: text,
    spellString: spellParts.join(' '),
    ttsString: ttsParts.join(' '),
  );
}

File: vi_spelling.py
"""
Vietnamese syllable decomposition and spelling string generator.

Provides functions to decompose Vietnamese syllables into components
(initial, nucleus, ending, tone) and generate "spelling song" strings
for TTS pronunciation.
"""

import unicodedata


def add_tone(syllable: str, tone: str) -> str:
    """
    Add Vietnamese tone mark to a syllable.

    Args:
        syllable: Base syllable without tone (e.g. "oc")
        tone: Tone code — 's' (sắc), 'f' (huyền), 'r' (nặng),
              'x' (hỏi), 'j' (ngã)

    Returns:
        Syllable with tone mark applied (e.g. "óc")
    """
    if not syllable or tone in (None, "", "0"):
        return syllable

    tone_map = {
        "s": "\u0301",  # sắc   (acute)
        "f": "\u0300",  # huyền (grave)
        "r": "\u0323",  # nặng  (dot below)
        "x": "\u0309",  # hỏi   (hook above)
        "j": "\u0303",  # ngã   (tilde)
    }

    combining = tone_map.get(tone.lower(), "")
    vowel_order = "aăâeêoôơiuưy"

    # Find best vowel position for tone placement
    best_pos = -1
    best_priority = len(vowel_order)

    for i, ch in enumerate(syllable):
        lower = ch.lower()
        if lower in vowel_order:
            priority = vowel_order.index(lower)
            if priority < best_priority:
                best_priority = priority
                best_pos = i

    if best_pos == -1:
        return syllable  # no vowel found

    return syllable[: best_pos + 1] + combining + syllable[best_pos + 1 :]


# Consonant lists used by the decomposer
INITIALS = [
    "ngh", "ng", "gh", "gi", "qu", "ch", "kh", "nh", "ph", "th", "tr",
    "b", "c", "d", "đ", "g", "h", "k", "l", "m", "n", "p", "q", "r",
    "s", "t", "v", "x",
]

ENDINGS = ["ch", "nh", "ng", "c", "m", "n", "p", "t"]

# Special cases for "gi" — common exceptions in Vietnamese primary schools
_GI_SPECIAL_CASES = {
    "gì": {
        "initial": "gi", "main": "", "ending": "", "tone": "Huyền",
        "spell": "gi huyền gì", "tts": "gi huyền gì",
    },
    "gìn": {
        "initial": "gi", "main": "in", "ending": "", "tone": "Huyền",
        "spell": "gi in gin huyền gìn", "tts": "gi in gin huyền gìn",
    },
    "giếng": {
        "initial": "gi", "main": "iêng", "ending": "", "tone": "Sắc",
        "spell": "gi iêng giêng sắc giếng", "tts": "gi iêng giêng sắc giếng",
    },
    "giềng": {
        "initial": "gi", "main": "iêng", "ending": "", "tone": "Huyền",
        "spell": "gi iêng giêng huyền giềng", "tts": "gi iêng giêng huyền giềng",
    },
    "giết": {
        "initial": "gi", "main": "iết", "ending": "", "tone": "Sắc",
        "spell": "gi iết giết sắc giết", "tts": "gi iết giết sắc giết",
    },
    "giêng": {
        "initial": "gi", "main": "iêng", "ending": "", "tone": "Ngang",
        "spell": "gi iêng giêng", "tts": "gi iêng giêng",
    },
}


def decompose_vietnamese_syllable(word: str) -> dict:
    """
    Decompose a Vietnamese syllable into its components and build
    spelling / TTS strings.

    Args:
        word: A single Vietnamese word/syllable.

    Returns:
        Dict with keys: initial, main, ending, tone, original,
        spell_string, tts_string
    """
    if not word:
        return {
            "initial": "", "main": "", "ending": "", "tone": "Ngang",
            "original": word, "spell_string": "", "tts_string": "",
        }

    # --- Check special cases first ---
    if word.lower() in _GI_SPECIAL_CASES:
        sp = _GI_SPECIAL_CASES[word.lower()]
        return {
            "initial": sp["initial"],
            "main": sp["main"] or "(none)",
            "ending": sp["ending"] or "(none)",
            "tone": sp["tone"],
            "original": word,
            "spell_string": sp["spell"],
            "tts_string": sp["tts"],
        }

    # --- Normalize to NFD and extract tone ---
    normalized = unicodedata.normalize("NFD", word.lower())

    tone_map = {
        "\u0300": "Huyền",
        "\u0309": "Hỏi",
        "\u0303": "Ngã",
        "\u0301": "Sắc",
        "\u0323": "Nặng",
    }

    tone = "Ngang"
    for char in normalized:
        if char in tone_map:
            tone = tone_map[char]
            break

    # Remove tone marks to get the base form
    base = "".join(ch for ch in normalized if ch not in tone_map)

    # --- Find initial consonant ---
    chars = list(base)
    initial = ""
    i = 0
    for init in sorted(INITIALS, key=len, reverse=True):
        if "".join(chars[: len(init)]) == init:
            initial = init
            i = len(init)
            break

    remaining = chars[i:]

    # --- Find ending consonant ---
    ending = ""
    if remaining:
        for end in sorted(ENDINGS, key=len, reverse=True):
            if "".join(remaining[-len(end) :]) == end:
                ending = end
                remaining = remaining[: -len(end)]
                break

    # --- Nucleus (main vowel) ---
    main = "".join(remaining)
    main_nfc = unicodedata.normalize("NFC", main)

    # --- Build spelling string ---
    parts = []

    # Handle "qu" special case
    if initial == "qu" and ending:
        main_nfc = "u" + main_nfc

    # 1. Spell main vowel letter by letter
    if len(main_nfc + ending) > 1:
        for letter in main_nfc:
            parts.append(letter)

    # 2. Spell ending
    if ending:
        parts.append(ending)

    # 3. Main + Ending together
    main_ending = main_nfc + ending
    if len(main_ending) > 1:
        parts.append(main_ending)

    # 4. Initial + Main_Ending
    syllable_no_tone = ""
    if initial:
        parts.append(initial)
        parts.append(main_ending)

        # Fix double-u for "qu"
        if initial == "qu" and ending:
            main_ending = main_ending[1:]

        syllable_no_tone = initial + main_ending
        if syllable_no_tone:
            parts.append(syllable_no_tone)

        # 5. Add tone name
        if tone != "Ngang":
            parts.append(tone.lower())

        # 6. Final word with tone
        if tone != "Ngang":
            parts.append(word)

    spell_string = " ".join(parts)

    # --- Build TTS-friendly string ---
    tts_parts = list(parts)  # copy
    for idx in range(len(tts_parts)):
        token = tts_parts[idx]

        # Add "ờ" to consonant clusters so TTS can pronounce them
        if token in INITIALS or token in ENDINGS:
            if token == "k":
                tts_parts[idx] = "ca"
            elif token == "ngh":
                tts_parts[idx] = "ngờ"
            elif token == "gh":
                tts_parts[idx] = "gờ"
            elif token != "gi":
                tts_parts[idx] = token + "ờ"

        # Add sắc tone to closed syllables ending in stop consonants
        if (token == main_ending or token == syllable_no_tone) and ending in [
            "ch", "c", "p", "t",
        ]:
            tts_parts[idx] = add_tone(tts_parts[idx], "s")

        # Special case: "qu" with stop ending
        if initial == "qu" and ending:
            if (
                len(tts_parts[idx]) > 1
                and tts_parts[idx][1:] == main_ending
                and ending in ["ch", "c", "p", "t"]
            ):
                tts_parts[idx] = add_tone(tts_parts[idx], "s")

    tts_string = " ".join(tts_parts)

    return {
        "initial": initial or "(none)",
        "main": main_nfc or "(none)",
        "ending": ending or "(none)",
        "tone": tone,
        "original": word,
        "spell_string": spell_string,
        "tts_string": tts_string,
    }


def get_spelling_for_text(text: str) -> str:
    """
    Generate a full TTS-readable spelling string for a Vietnamese sentence.

    Splits the text into words, decomposes each, and joins their
    tts_string values with pauses (periods).

    Args:
        text: Vietnamese sentence.

    Returns:
        TTS-friendly spelling string for the whole sentence.
    """
    import re

    words = re.findall(r"[\w]+", text, re.UNICODE)
    tts_parts = []
    for w in words:
        result = decompose_vietnamese_syllable(w)
        if result["tts_string"]:
            tts_parts.append(result["tts_string"])
    return ". ".join(tts_parts)